File: otto-android/app/src/main/java/to/ottomot/driftd/MapDriveHorizonDepth.kt
package to.ottomot.driftd

import kotlin.math.roundToInt

/** Map-proportional marker scale and distance-based overlap priority during pitched drive follow. */
internal object MapDriveHorizonDepth {
    /** Checkpoints farther than this from the user are not shown on the map route layer. */
    const val CHECKPOINT_VISIBLE_MAX_DISTANCE_METERS = 1609.344

    const val ROUTE_MIN_SCALE = 0.55f
    const val PRESENCE_MIN_SCALE = 0.5f

    private const val SCALE_QUANTIZATION_STEP = 0.05f
    private const val ENDPOINT_PRIORITY_BOOST = 200_000_000L
    private const val PRESENCE_PRIORITY_BOOST = 150_000_000L
    private const val PRIORITY_BASE = 1_000_000L

    fun visibleMapHeightMeters(latitudeDelta: Double): Double {
        if (!latitudeDelta.isFinite() || latitudeDelta <= 0) return 111_000.0
        return maxOf(50.0, latitudeDelta * 111_000.0)
    }

    fun horizonScale(
        distanceMeters: Double,
        visibleMapHeightMeters: Double,
        minScale: Float = ROUTE_MIN_SCALE,
    ): Float {
        if (!distanceMeters.isFinite() || !visibleMapHeightMeters.isFinite() || visibleMapHeightMeters <= 0) {
            return 1f
        }
        val t = (distanceMeters / visibleMapHeightMeters).coerceIn(0.0, 1.0)
        val minS = minScale.toDouble()
        val raw = maxOf(minS, 1.0 - t * (1.0 - minS))
        return quantizeScale(raw.toFloat())
    }

    fun driveRouteOverlapPriority(
        distanceMeters: Double,
        markerType: String? = null,
        tieBreaker: Int = 0,
    ): Long {
        if (!distanceMeters.isFinite()) return tieBreaker.toLong()
        var priority = PRIORITY_BASE - distanceMeters.roundToInt() + tieBreaker
        if (markerType == "start" || markerType == "finish") {
            priority += ENDPOINT_PRIORITY_BOOST
        }
        return priority
    }

    fun drivePresenceOverlapPriority(
        distanceMeters: Double,
        tieBreaker: Int = 0,
    ): Long = driveRouteOverlapPriority(distanceMeters, tieBreaker = tieBreaker) + PRESENCE_PRIORITY_BOOST

    /** Waypoints only; start/finish/stop are always eligible for display. */
    fun shouldShowRouteMarker(
        markerType: String?,
        distanceMeters: Double?,
    ): Boolean {
        if (markerType != "waypoint") return true
        if (distanceMeters == null || !distanceMeters.isFinite()) return true
        return distanceMeters <= CHECKPOINT_VISIBLE_MAX_DISTANCE_METERS
    }

    /** Presence pins during pitched drive follow — same ~1 mi horizon as route checkpoints. */
    fun shouldShowPresenceMarker(distanceMeters: Double?): Boolean {
        if (distanceMeters == null || !distanceMeters.isFinite()) return true
        return distanceMeters <= CHECKPOINT_VISIBLE_MAX_DISTANCE_METERS
    }

    private fun quantizeScale(scale: Float): Float {
        val step = SCALE_QUANTIZATION_STEP
        return (scale / step).roundToInt() * step
    }
}

File: otto-mobile/otto-mobile/MapDriveHorizonDepth.swift
import CoreLocation
import Foundation

/// Map-proportional marker scale and distance-based overlap priority during pitched drive follow.
enum MapDriveHorizonDepth {
    /// Checkpoints farther than this from the user are not shown on the map tab route layer.
    static let checkpointVisibleMaxDistanceMeters: Double = 1609.344

    static let routeMinScale: CGFloat = 0.55
    static let presenceMinScale: CGFloat = 0.50
    private static let scaleQuantizationStep: CGFloat = 0.05
    private static let endpointPriorityBoost = 200_000_000
    private static let presencePriorityBoost = 150_000_000
    private static let priorityBase = 1_000_000

    /// Visible north–south span in meters from the same latitude-delta signal as marker LOD.
    static func visibleMapHeightMeters(latitudeDelta: Double) -> Double {
        guard latitudeDelta.isFinite, latitudeDelta > 0 else { return 111_000 }
        return max(50, latitudeDelta * 111_000)
    }

    static func horizonScale(
        distanceMeters: Double,
        visibleMapHeightMeters: Double,
        minScale: CGFloat = routeMinScale
    ) -> CGFloat {
        guard distanceMeters.isFinite, visibleMapHeightMeters.isFinite, visibleMapHeightMeters > 0 else {
            return 1
        }
        let t = min(1, max(0, distanceMeters / visibleMapHeightMeters))
        let minS = Double(minScale)
        let raw = max(minS, 1.0 - t * (1.0 - minS))
        return quantizeScale(CGFloat(raw))
    }

    static func driveRouteOverlapPriority(
        distanceMeters: Double,
        markerType: String?,
        tieBreaker: Int = 0
    ) -> Int {
        guard distanceMeters.isFinite else { return tieBreaker }
        var priority = priorityBase - Int(distanceMeters.rounded()) + tieBreaker
        if markerType == "start" || markerType == "finish" {
            priority += endpointPriorityBoost
        }
        return priority
    }

    static func drivePresenceOverlapPriority(
        distanceMeters: Double,
        tieBreaker: Int = 0
    ) -> Int {
        driveRouteOverlapPriority(
            distanceMeters: distanceMeters,
            markerType: nil,
            tieBreaker: tieBreaker
        ) + presencePriorityBoost
    }

    static func distanceMeters(
        from user: CLLocation,
        to coordinate: CLLocationCoordinate2D
    ) -> Double {
        user.distance(
            from: CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)
        )
    }

    /// Waypoints only; start/finish/stop are always eligible for display.
    static func shouldShowRouteMarker(
        markerType: String?,
        distanceMeters: Double?
    ) -> Bool {
        guard markerType == "waypoint" else { return true }
        guard let distanceMeters, distanceMeters.isFinite else { return true }
        return distanceMeters <= checkpointVisibleMaxDistanceMeters
    }

    /// Presence pins during pitched drive follow — same ~1 mi horizon as route checkpoints.
    static func shouldShowPresenceMarker(distanceMeters: Double?) -> Bool {
        guard let distanceMeters, distanceMeters.isFinite else { return true }
        return distanceMeters <= checkpointVisibleMaxDistanceMeters
    }

    private static func quantizeScale(_ scale: CGFloat) -> CGFloat {
        let step = scaleQuantizationStep
        return (scale / step).rounded() * step
    }
}
